src/bytecode/mod.rs

```diff
@@ -1,27 +1,37 @@
 pub mod instructions;
 
 use std::vec::Vec;
+use std::collections::HashMap;
 use std::fmt;
-use self::instructions::Instr;
+use self::instructions::{Reg, Instr, Value};
 
 /// A simpler representation of Lua
 pub struct LuaBytecode {
     block: Vec<Instr>,
-    reg_counter: u64
+    reg_counter: u64,
+    registers: Vec<Reg>,
+    reg_map: HashMap<String, u64>
 }
 
-impl<'a> LuaBytecode {
+impl LuaBytecode {
     pub fn new() -> LuaBytecode {
         LuaBytecode {
             block: vec![],
-            reg_counter: 0
+            reg_counter: 0,
+            registers: vec![],
+            reg_map: HashMap::new()
         }
     }
 
+    /// Get the number of instructions that are part of this block.
+    pub fn get_num_of_instrs(&self) -> usize {
+        self.block.len()
+    }
+
     /// Get the list of instructions that can be executed in order
     /// to perform some computation.
-    pub fn get_instrs(&'a self) -> &'a Vec<Instr> {
-        &self.block
+    pub fn get_instr(&self, index: usize) -> Instr {
+        self.block[index].clone()
```

Comment:
Why are we cloning the instruction? Could we not return a reference? `get_instr(...) -> &Instr` and avoid the clone cost?

Reply:
OK, let's do that in a later PR. I agree that we don't want to do this in long term, but if we make this PR perfect, we'll never get to a second PR!

Reply:
That's true!